scipy/sparse/linalg/dsolve/linsolve.py

```diff
@@ -383,3 +384,125 @@ def solve(b):
         return solve
     else:
         return splu(A).solve
+
+
+def spsolve_triangular(A, b, lower=True, overwrite_b=False):
+    """
+    Solve the equation `A x = b` for `x`, assuming a is a triangular matrix.
+    Parameters
+    ----------
+    A : (M, M) sparse matrix
+        A sparse square triangular matrix. Should be in CSR format.
+    b : (M,) or (M, N) array_like
+        Right-hand side matrix in `a x = b`
+    lower : bool, optional
+        Whether `A` is a lower or upper triangle matrix.
+        Default is lower triangle matrix.
+    overwrite_b : bool, optional
+        Allow overwriting data in `b`.
+        This may give a performance gain.
+    Returns
+    -------
+    x : (M,) or (M, N) ndarray
+        Solution to the system `A x = b`.  Shape of return matches shape of `b`.
+    Raises
+    ------
+    LinAlgError
+        If `A` is singular.
+    ValueError
+        If shape of `A` or shape of `b` do not match the requirements.
+    Notes
+    -----
+    .. versionadded:: 0.20.0
+    """
+
+    ## check input
+    if A.ndim != 2:
+        raise ValueError('A must have 2 dimensions but its shape is {}.'.format(A.shape))
+    if A.shape[0] != A.shape[1]:
+        raise ValueError('A must be a square matrix but its shape is {}.'.format(A.shape))
+
+    if not isspmatrix_csr(A):
+        warn('CSR matrix format is required. Converting to CSR matrix.',
+             SparseEfficiencyWarning)
+        A = csr_matrix(A)
+
+    A.eliminate_zeros()
+    A.sort_indices()
+
+    b = np.asanyarray(b)
+
+    if b.ndim not in [1, 2]:
+        raise ValueError('b must have 1 or 2 dims but its shape is {}.'.format(b.shape))
+    if A.shape[0] != b.shape[0]:
+        raise ValueError('The size of the dimensions of A must be equal to '
+            'the size of the first dimension of b but the shape of A is '
+            '{} and the shape of b is {}.'.format(A.shape, b.shape))
+
+    ## init x
+    if not overwrite_b:
+        x = np.empty(b.shape)
+    else:
+        x = b
+
+    ## fill x
+    if lower:
+
+        ## fill forward
+        indptr_start = A.indptr[0]
+        for i in range(len(b)):
+            indptr_stop = A.indptr[i+1]
+
+            ## check regularity and triangularity
+            assert np.all(A.indices[indptr_start:indptr_stop-1] <
+                A.indices[indptr_start+1:indptr_stop])     # indices are assumed to be sorted
+            if indptr_stop <= indptr_start or A.indices[indptr_stop-1] < i:
+                raise LinAlgError('A is singular since the {}th diagonal '
+                    'entry of A is zero!'.format(i))
+            if A.indices[indptr_stop-1] > i:
+                raise LinAlgError('A is no lower triangle matrix, since the '
+                    'entry at ({},{}) is not zero!'.format(i, A.indices[indptr_stop-1]))
```

Comment:
These error messages need some revision:
 * "A is singular: {}th diagonal is zero"
 * "A is not lower triangular: entry [{},{}] is nonzero"